### DISCUSSION TITLE
Preparing new version 0.1

### DISCUSSION THREAD

**Discussion by Kalmat** [Kalmat/PyWinCtl#60] [Announcements]
Hi all,

I'm working (mainly trying to figure out how to deeply test everything) on a new version (v0.1) which is a major update with a number of new features, improvements and fixes:

- ALL: 
    - Experimental Multi-Monitor support (macOS tests still pending), including a new general function (monitorPlugDetection()) to instantly detect changes in multi-monitor configuration (dynamically detects monitors plugged to or unplugged from system)
    - Replaced PyRect by custom class (PyRect has no multi-monitor support)
    - Added bbox and rect properties (x, y, right, bottom)
    - Added setParent() method
    - Improved requirements.txt and setup.py

- LINUX:
    - Replaced ewmh module by custom class (ewmh was showing its "age") - Thanks to [MestreLion](https://github.com/MestreLion)
    - Removed pyinput dependency (found a better solution for sendBehind() in Linux Mint)
    - Improved activate() and restore() methods to better comply with EWMH specs (and properly work on Unity) - Thanks to [MestreLion](https://github.com/MestreLion)
    - Fixed acceptInput()
    - Fixed move and resize (gravity and flags) - Thanks to [elraymond](https://github.com/elraymond)

- WINDOWS:
     - getAllWindows() and related return "user" apps only
     - Simplified alwaysOnBottom()
     - Found a smarter way to refresh window after bringing it back with sendBehind()

- MACOSWINDOW:
    - Fixed getClientFrame() to be invoked from non-main threads - Thanks to [super-ibby](https://github.com/super-ibby)

- MACOSNSWINDOW:
    - Added experimental acceptInput() method (not possible in AppleScript version)

- MACOS:
    - Added 'unflipValues' param to getMousePos() function to convert mouse position to flipped/unflipped screen coordinates

This new version will require Python 3.7+, mainly because of typing requirements (like annotations). [Personal reflection: perhaps it is because of my limited typing skills, but it's very weird to me that typing may condition "real" functionalities or compatibility]

TESTS
Since this new version is a major change, the risk of hidden (or not so hidden) issues arising is high. I tried to test everything as much as possible, but I'm affraid "real life" is more complex and diverse than anything you can imagine upfront. Once released, if any current/new user finds any problem with this new version, please open an issue to let me know as soon as possible.

I couldn't test in a macOS multi-monitor setup. If anyone can help, he is more than welcome. Besides, if someone knows how to enable multiple monitors in a macOS virtualbox VM, I would really appreciate if he can point me to a workaround.

COMMENTS, REQUESTS AND SUGGESTIONS
We are not in any hurry, so there is time enough to include any comment or suggestion you may have. Just let me know and we will analyse if it can be included in this new version before it's released.

**Comment by WantToLearnJapanese** [comment]
I am so excited since it seems PyWinCtl could become the base library for Multi-platform supported window manager.

**Comment by Kalmat** [comment]
Hi again!

Sorry to be so late with new version 0.1. The delay is because of two reasons:

1. I have been working on two new modules, in which PyWinCtl relies (or will rely) on:

    - PyWinBox, replacing PyRect in order to (hopefully) add multi-monitor support.
    - PyMonCtl, which includes a huge number of features to get info on and control monitors plugged (of course, also very helpful for multi-monitor setups)

2. I am trying to get an actual macOS to test everything, and most especially, all multi-monitor features from both new modules. I didn't want to release the new version until all this was done.

In the meantime, with the aim of fixing some recent issues sent by several users, I have released a new "intermediate" version: v0.0.45, which inevitably includes many of the features intended to be included within PyWinCtl-0.1:

- ALL:
    - Multi-Monitor support thru PyWinBox module (macOS tests still pending)
    - Added bbox (x, y, width, height) and rect properties (x, y, right, bottom)
    - Added setParent() method
    - Improved requirements.txt and setup.py

- LINUX:

    - Replaced ewmh module by custom class (ewmh was showing its "age") - Thanks to [MestreLion](https://github.com/MestreLion).
    - Removed pynput dependency (found a better solution for sendBehind() in Linux Mint)
    - Improved activate() and restore() methods to better comply with EWMH specs (and properly work on Unity) - Thanks to [MestreLion](https://github.com/MestreLion)
    - Fixed acceptInput()
    - Fixed move and resize (gravity and flags) - Thanks to [elraymond](https://github.com/elraymond)

- WINDOWS:

    - Rolled back to previous getAllWindows() implementation since the new one (aimed to return "user" apps only) was missing relevant windows
    - Simplified alwaysOnBottom()
    - Found a smarter way to refresh window after bringing it back with sendBehind()

- MACOSWINDOW:

    - Fixed getClientFrame() to be invoked from non-main threads - Thanks to [super-ibby](https://github.com/super-ibby)

- MACOSNSWINDOW:

    - Added experimental acceptInput() method (not possible in AppleScript version)

- MACOS:

    - Added 'unflipValues' param to convert between flipped/unflipped screen coordinates in all related functions (e.g. window box properties, getMousePos(), etc.)


In short, new v0.1 version will include new PyMonCtl module and, hopefully, it will be fully tested in all actual platforms, especially intensive multi-monitor tests!

Keep pushing, guys!
